chirp/drivers/tdh8.py

```diff
@@ -0,0 +1,2222 @@
+            break
+
+    if len(response) in [8, 12]:
+        # DEBUG
+        LOG.info("Valid response, got this:")
+        LOG.info(util.hexprint(response))
+        if len(response) == 12:
+            ident = response[0] + response[3] + response[5] + response[7:]
+        else:
+            ident = response
+    else:
+        # bad response
+        msg = "Unexpected response, got this:"
+        msg += util.hexprint(response)
+        LOG.debug(msg)
+        raise errors.RadioError("Unexpected response from radio.")
+
+    if secondack:
+        serial.write(b"\x06")
+        ack = serial.read(1)
+        if ack != b"\x06":
+            raise errors.RadioError("Radio refused clone")
+
+    return ident
+
+
+def response_mode(mode):
+    data = mode
+    return data
+
+
+def _read_block(radio, start, size):
+    serial = radio.pipe
+
+    cmd = struct.pack(">cHb", b'R', start, size)
+    expectedresponse = b"W" + cmd[1:]
+
+    try:
+        serial.write(cmd)
+        response = serial.read(5 + size)
+        if response[:4] != expectedresponse:
+            raise Exception("Error reading block %04x." % (start))
+
+        block_data = response[4:-1]
+        # print("block_data = {}".format(str(binascii.b2a_hex(block_data))))
+
+    except:
+        raise errors.RadioError("Failed to read block at %04x" % start)
+
+    return block_data
+
+
+def _get_radio_firmware_version(radio):
+    if radio.MODEL in ALL_MODEL:
+        block = _read_block(radio, 0x1B40, 0x20)
+        version = block[0:6]
+    return version
+
+
+IDENT_BLACKLIST = {
+    b"\x50\x56\x4F\x4A\x48\x1C\x14": "Radio identifies as TIDRADIO TD-H8",
+}
+
+
+def _ident_radio(radio):
+    for magic in radio._idents:
+        error = None
+        try:
+            data = _do_ident(radio, magic)
+            return data
+        except errors.RadioError as e:
+            error = e
+            time.sleep(2)
+
+    if error:
+        raise error
+    raise errors.RadioError("Radio did not respond")
+
+
+def _do_download(radio):
+    data = _ident_radio(radio)
+    append_model = False
+    # HAM OR GMRS
+    # Determine the walkie-talkie mode
+    # TDH8 have three mode:ham, gmrs and normal
+    radio.radio_ident = str(data)[2:8]
+    LOG.info("Radio mode is " + radio.radio_ident)
+    LOG.info("Chirp choose mode is " + radio.ident_Mode)
+    # The Ham and Gmrs modes are subclasses of this model tdh8.
+    # When the Ham or Gmrs mode is selected,
+    # a double verification operation will be performed to
+    # obtain the current mode code of the walkie-talkie and
+    # the mode code currently selected by the chirp for comparison.
+    # If the two are consistent, enter This mode,
+    # otherwise an inconsistent error will be prompted.
+    if radio.radio_ident == radio.ident_Mode:
+        LOG.info("Successful match.")
+    else:
+        msg = (
+            "Please check whether the walkie-talkie mode is consistent with "
+            "the selected model mode.")
```

Comment:
Our more conventional error message for this would be something like *Unable to talk to this model* or *Model mismatch*. Both "walkie talkie" and "model mode" do not sound very good to native english speakers.